pl_bolts/models/gans/pix2pix/components.py

```diff
@@ -0,0 +1,167 @@
+import torch
+from torch import nn
+
+
+def center_crop(image, new_shape):
+    h, w = image.shape[-2:]
+    n_h, n_w = new_shape[-2:]
+    cy, cx = int(h / 2), int(w / 2)
+    xmin, ymin = cx - n_w // 2, cy - n_h // 2
+    xmax, ymax = xmin + n_w, ymin + n_h
+    cropped_image = image[..., xmin:xmax, ymin:ymax]
+    return cropped_image
+
+
+class ConvBlock(nn.Module):
+    def __init__(self, in_channels, out_channels, use_dropout=False, use_bn=True):
+        super(ConvBlock, self).__init__()
+        self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1)
+        self.activation = nn.LeakyReLU(0.2)
+
+        if use_bn:
+            self.batchnorm = nn.BatchNorm2d(out_channels)
+        self.use_bn = use_bn
+
+        if use_dropout:
+            self.dropout = nn.Dropout()
+        self.use_dropout = use_dropout
+
+    def forward(self, x):
+        x = self.conv1(x)
+        if self.use_bn:
+            x = self.batchnorm(x)
+        if self.use_dropout:
+            x = self.dropout(x)
+        x = self.activation(x)
+        return x
+
+
+class UpSampleConv(nn.Module):
+
+    def __init__(self, input_channels, use_dropout=False, use_bn=True):
+        super(self).__init__()
+        self.upsample = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
```

Comment:
Let's just confirm once if Upsample is done using `nn.Upsample` or `nn.ConvTranspose2d` both work fine. I haven't read Pix2Pix paper so let me check once.

Reply:
Hi, thank you for the review. In section 6 of the Pix2Pix paper authors have mentioned that they upsampled the tensors by a factor of 2 but they haven't exactly mentioned if Transposed Conv is used or Upsample followed by Conv layer.

Reply:
I quickly checked [the paper](https://arxiv.org/pdf/1611.07004.pdf) and found that [the PyTorch implementation](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/f13aab8148bd5f15b9eb47b690496df8dadbab0c/models/networks.py#L315) linked from [the author's Lua implementation](https://github.com/phillipi/pix2pix) uses `nn.ConvTranspose2d`, so shall we follow that architecture unless someone has a strong opinion?

Reply:
Yes I too confirmed that it is `nn.ConvTranspose2d`. I have referred [TensorFlow docs,](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/generative/pix2pix.ipynb#scrollTo=nhgDsHClSQzP) which give a really nice implementation.